From 6554b24d3dab3cf4ef2ea4cdb35a0ae3cf2a910e Mon Sep 17 00:00:00 2001
From: Jian-Hong Pan <jhp@endlessos.org>
Date: Tue, 17 Jun 2025 16:37:25 +0800
Subject: [PATCH 4/6] [Endless] sd-boot: Read fake symlinks

ostree uses symlinks on the boot filesystem, this isn't great when that
filesystem is vfat, which can't do symlinks. Since the EFI ESP must be
vfat on most implementation, this makes ostree incompaible with sd-boot.

To allow ostree to keep making symlinks, we make a fake symlink that's
just a text file with the name of the file that would be linked to.

AC: EOS uning sd-boot as the boot loader can boot with OSTree directly.

https://phabricator.endlessm.com/T27040
https://phabricator.endlessm.com/T35070
---
 ...4-Endless-sd-boot-Read-fake-symlinks.patch | 116 ++++++++++++++++++
 1 file changed, 116 insertions(+)
 create mode 100644 patches/systemd/0004-Endless-sd-boot-Read-fake-symlinks.patch

diff --git a/patches/systemd/0004-Endless-sd-boot-Read-fake-symlinks.patch b/patches/systemd/0004-Endless-sd-boot-Read-fake-symlinks.patch
new file mode 100644
index 000000000..50e12c94c
--- /dev/null
+++ b/patches/systemd/0004-Endless-sd-boot-Read-fake-symlinks.patch
@@ -0,0 +1,116 @@
+From 25e0742724fa2e6ca9c87b88cc16845a8581d47c Mon Sep 17 00:00:00 2001
+From: Derek Foreman <derek@endlessm.com>
+Date: Tue, 27 Feb 2024 18:26:50 +0800
+Subject: [PATCH 4/5] [Endless] sd-boot: Read fake symlinks
+
+ostree uses symlinks on the boot filesystem, this isn't great when that
+filesystem is vfat, which can't do symlinks. Since the EFI ESP must be
+vfat on most implementation, this makes ostree incompaible with sd-boot.
+
+To allow ostree to keep making symlinks, we make a fake symlink that's
+just a text file with the name of the file that would be linked to.
+
+AC: EOS uning sd-boot as the boot loader can boot with OSTree directly.
+
+https://phabricator.endlessm.com/T27040
+https://phabricator.endlessm.com/T35070
+---
+ src/boot/efi/boot.c | 57 ++++++++++++++++++++++++++++++++++-----------
+ 1 file changed, 43 insertions(+), 14 deletions(-)
+
+diff --git a/src/boot/efi/boot.c b/src/boot/efi/boot.c
+index 5e6f142dd7..d4d45d443c 100644
+--- a/src/boot/efi/boot.c
++++ b/src/boot/efi/boot.c
+@@ -1554,9 +1554,43 @@ static EFI_STATUS efivar_get_timeout(const char16_t *var, uint64_t *ret_value) {
+         return EFI_SUCCESS;
+ }
+ 
++static char16_t *resolve_link(
++                EFI_FILE *root_dir,
++                const char16_t *link,
++                const char16_t *file) {
++        EFI_STATUS err;
++        _cleanup_free_ char *contents = NULL;
++        _cleanup_free_ char16_t *linkname = NULL;
++        _cleanup_free_ char16_t *target = NULL;
++        char16_t *out = NULL;
++
++        err = BS->AllocatePool(EfiReservedMemoryType, strsize16(link) + strsize16(L".sln"), (void **)&linkname);
++        if (err != EFI_SUCCESS)
++                return NULL;
++        strcpy16(linkname, link);
++        strcpy16(linkname + strlen16(linkname), L".sln");
++
++        err = file_read(root_dir, linkname, 0, 0, &contents, NULL);
++        if (err != EFI_SUCCESS)
++                return NULL;
++
++        target = xstr8_to_path(contents);
++        if (!target)
++                return NULL;
++
++        err = BS->AllocatePool(EfiReservedMemoryType, strsize16(target) + strsize16(file), (void **)&out);
++        if (err != EFI_SUCCESS)
++                return NULL;
++        strcpy16(out, target);
++        strcpy16(out + strlen16(out), file);
++
++        return out;
++}
++
+ static void config_load_defaults(Config *config, EFI_FILE *root_dir) {
+         _cleanup_free_ char *content = NULL;
+         size_t content_size, value = 0;  /* avoid false maybe-uninitialized warning */
++        _cleanup_free_ char16_t *link = NULL;
+         EFI_STATUS err;
+ 
+         assert(root_dir);
+@@ -1573,20 +1607,11 @@ static void config_load_defaults(Config *config, EFI_FILE *root_dir) {
+                 .timeout_sec_efivar = TIMEOUT_UNSET,
+         };
+ 
+-        err = file_read(root_dir, u"\\loader\\loader.conf", 0, 0, &content, &content_size);
++        link = resolve_link(root_dir, L"\\loader", L"\\entries");
++        if (!link)
++                link = L"\\loader\\entries";
++        err = file_read(root_dir, link, 0, 0, &content, &content_size);
+         if (err == EFI_SUCCESS) {
+-                /* First, measure. */
+-                err = tpm_log_tagged_event(
+-                                TPM2_PCR_BOOT_LOADER_CONFIG,
+-                                POINTER_TO_PHYSICAL_ADDRESS(content),
+-                                content_size,
+-                                LOADER_CONF_EVENT_TAG_ID,
+-                                u"loader.conf",
+-                                /* ret_measured= */ NULL);
+-                if (err != EFI_SUCCESS)
+-                        log_error_status(err, "Error measuring loader.conf into TPM: %m");
+-
+-                /* Then: parse */
+                 config_defaults_load_from_file(config, content);
+         }
+ 
+@@ -1637,6 +1662,7 @@ static void config_load_type1_entries(
+         _cleanup_free_ EFI_FILE_INFO *f = NULL;
+         size_t f_size = 0;
+         EFI_STATUS err;
++        _cleanup_free_ char16_t *link = NULL;
+ 
+         assert(config);
+         assert(device);
+@@ -1644,7 +1670,10 @@ static void config_load_type1_entries(
+ 
+         /* Adds Boot Loader Type #1 entries (i.e. /loader/entries/….conf) */
+ 
+-        err = open_directory(root_dir, u"\\loader\\entries", &entries_dir);
++        link = resolve_link(root_dir, L"\\loader", L"\\entries");
++        if (!link)
++                link = L"\\loader\\entries";
++        err = open_directory(root_dir, link, &entries_dir);
+         if (err != EFI_SUCCESS)
+                 return;
+ 
+-- 
+2.49.0
+
-- 
2.49.0

